include/envoy/network/io_handle.h

```diff
@@ -1,33 +1,136 @@
 #pragma once
 
+#include <bitset>
 #include <memory>
 
 #include "envoy/common/pure.h"
 
 namespace Envoy {
+
+namespace Buffer {
+struct RawSlice;
+}
+
 namespace Network {
 
+namespace Address {
+class Instance;
+} // namespace Address
+
+class IoError {
+public:
+  enum class IoErrorCode {
+    NoError = 0,
+    Again,
+    NoSupport,
+    AddressFamilyNoSupport,
+    InProgress,
+    Permission,
+    UnknownError
+  };
+  virtual ~IoError() {}
+
+  // Map platform specific error into IoErrorCode.
+  virtual IoErrorCode getErrorCode() const PURE;
+
+  virtual std::string getErrorDetails() const PURE;
+};
+
+/**
+ * Basic type for return result which has a return code and error code defined
+ * according to different implementations.
+ */
+template <typename T> struct IoHandleCallResult {
+  virtual ~IoHandleCallResult() {}
+
+  T rc_;
+  std::unique_ptr<IoError> err_;
+};
+
+using IoHandleCallIntResult = IoHandleCallResult<int>;
+using IoHandleCallSizeResult = IoHandleCallResult<ssize_t>;
+
 /**
  * IoHandle: an abstract interface for all I/O operations
  */
 class IoHandle {
 public:
-  IoHandle() {}
+  enum class ShutdownType { Read = 0, Write, Both };
+
+  enum class IoHandleFlag { NonBlock = 1, Append = 2 };
```

Comment:
nit: Why explicitly number this?

Reply:
...and ditto for the enums above (ShutdownType, IOError, ...).

Reply:
OK, can you use `0x1`, `0x10`, etc. do make that clear?

Reply:
I assume you meant 0b1, 0b10. Done

Reply:
Oops sorry I meant `0x1, 0x2, 0x4, 0x10`, etc. but either way.